src/PSM3.sol

```diff
@@ -154,25 +177,16 @@ contract PSM3 is IPSM3 {
     /*** Swap preview functions                                                                 ***/
     /**********************************************************************************************/
 
-    function previewSwap(address assetIn, address assetOut, uint256 amountIn)
+    function previewSwapExactIn(address assetIn, address assetOut, uint256 amountIn)
         public view override returns (uint256 amountOut)
     {
-        if (assetIn == address(asset0)) {
-            if      (assetOut == address(asset1)) return _previewOneToOneSwap(amountIn, _asset0Precision, _asset1Precision);
-            else if (assetOut == address(asset2)) return _previewSwapToAsset2(amountIn, _asset0Precision);
-        }
-
-        else if (assetIn == address(asset1)) {
-            if      (assetOut == address(asset0)) return _previewOneToOneSwap(amountIn, _asset1Precision, _asset0Precision);
-            else if (assetOut == address(asset2)) return _previewSwapToAsset2(amountIn, _asset1Precision);
-        }
-
-        else if (assetIn == address(asset2)) {
-            if      (assetOut == address(asset0)) return _previewSwapFromAsset2(amountIn, _asset0Precision);
-            else if (assetOut == address(asset1)) return _previewSwapFromAsset2(amountIn, _asset1Precision);
-        }
+        amountOut = _getSwapQuote(assetIn, assetOut, amountIn);
+    }
 
-        revert("PSM3/invalid-asset");
+    function previewSwapExactOut(address assetIn, address assetOut, uint256 amountOut)
+        public view override returns (uint256 amountIn)
+    {
+        amountIn = _getSwapQuote(assetOut, assetIn, amountOut);
```

Comment:
This needs to round up for the division. You should not be able to receive `amountOut` value greater than the `amountIn`. Maybe pass a bool for round down or up?

Reply:
All rounding behaviour is included in here: https://github.com/marsfoundation/spark-psm/pull/22

This is just basically making very similar logic to other swap to make easy review

Reply:
Ah ok